### ISSUE TITLE
How to train a multi task detection?

### ISSUE THREAD

**Issue by justStarG** [ultralytics/yolov5#4690]
## ❔Question
How to train a yolo model with different heads so that I can detect for different tasks ?
## Additional context
I want to use yolo to detect document paper from pictures, and detect many areas, such as  text area, figure area, title area and so on. But document box should be only one and much larger than others. So I want to use diffent detection heads but share the same backbone. How to implement this ? Should I define two head in `yaml` like:

```
backbone:
  # [from, number, module, args]
  [[-1, 1, Focus, [64, 3]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, BottleneckCSP, [128]],
   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
   [-1, 9, BottleneckCSP, [256]],
   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
   [-1, 9, BottleneckCSP, [512]],
   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
   [-1, 1, SPP, [1024, [5, 9, 13]]],
   [-1, 3, BottleneckCSP, [1024, False]],  # 9
  ]

# YOLOv5 head
head1:
  [[-1, 1, Conv, [512, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 6], 1, Concat, [1]],  # cat backbone P4
   [-1, 3, BottleneckCSP, [512, False]],  # 13

   [-1, 1, Conv, [256, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 4], 1, Concat, [1]],  # cat backbone P3
   [-1, 3, BottleneckCSP, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 14], 1, Concat, [1]],  # cat head P4
   [-1, 3, BottleneckCSP, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, BottleneckCSP, [1024, False]],  # 23 (P5/32-large)

   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  ]

head2:
  [[-1, 1, Conv, [512, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 6], 1, Concat, [1]],  # cat backbone P4
   [-1, 3, BottleneckCSP, [512, False]],  # 13

   [-1, 1, Conv, [256, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 4], 1, Concat, [1]],  # cat backbone P3
   [-1, 3, BottleneckCSP, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 14], 1, Concat, [1]],  # cat head P4
   [-1, 3, BottleneckCSP, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, BottleneckCSP, [1024, False]],  # 23 (P5/32-large)

   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  ]
```

**Comment by ahong007007** [comment]
> ## ❔Question
> How to train a yolo model with different heads so that I can detect for different tasks ?
> 
> ## Additional context
> I want to use yolo to detect document paper from pictures, and detect many areas, such as text area, figure area, title area and so on. But document box should be only one and much larger than others. So I want to use diffent detection heads but share the same backbone. How to implement this ? Should I define two head in `yaml` like:
> 
> ```
> backbone:
>   # [from, number, module, args]
>   [[-1, 1, Focus, [64, 3]],  # 0-P1/2
>    [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
>    [-1, 3, BottleneckCSP, [128]],
>    [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
>    [-1, 9, BottleneckCSP, [256]],
>    [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
>    [-1, 9, BottleneckCSP, [512]],
>    [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
>    [-1, 1, SPP, [1024, [5, 9, 13]]],
>    [-1, 3, BottleneckCSP, [1024, False]],  # 9
>   ]
> 
> # YOLOv5 head
> head1:
>   [[-1, 1, Conv, [512, 1, 1]],
>    [-1, 1, nn.Upsample, [None, 2, 'nearest']],
>    [[-1, 6], 1, Concat, [1]],  # cat backbone P4
>    [-1, 3, BottleneckCSP, [512, False]],  # 13
> 
>    [-1, 1, Conv, [256, 1, 1]],
>    [-1, 1, nn.Upsample, [None, 2, 'nearest']],
>    [[-1, 4], 1, Concat, [1]],  # cat backbone P3
>    [-1, 3, BottleneckCSP, [256, False]],  # 17 (P3/8-small)
> 
>    [-1, 1, Conv, [256, 3, 2]],
>    [[-1, 14], 1, Concat, [1]],  # cat head P4
>    [-1, 3, BottleneckCSP, [512, False]],  # 20 (P4/16-medium)
> 
>    [-1, 1, Conv, [512, 3, 2]],
>    [[-1, 10], 1, Concat, [1]],  # cat head P5
>    [-1, 3, BottleneckCSP, [1024, False]],  # 23 (P5/32-large)
> 
>    [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
>   ]
> 
> head2:
>   [[-1, 1, Conv, [512, 1, 1]],
>    [-1, 1, nn.Upsample, [None, 2, 'nearest']],
>    [[-1, 6], 1, Concat, [1]],  # cat backbone P4
>    [-1, 3, BottleneckCSP, [512, False]],  # 13
> 
>    [-1, 1, Conv, [256, 1, 1]],
>    [-1, 1, nn.Upsample, [None, 2, 'nearest']],
>    [[-1, 4], 1, Concat, [1]],  # cat backbone P3
>    [-1, 3, BottleneckCSP, [256, False]],  # 17 (P3/8-small)
> 
>    [-1, 1, Conv, [256, 3, 2]],
>    [[-1, 14], 1, Concat, [1]],  # cat head P4
>    [-1, 3, BottleneckCSP, [512, False]],  # 20 (P4/16-medium)
> 
>    [-1, 1, Conv, [512, 3, 2]],
>    [[-1, 10], 1, Concat, [1]],  # cat head P5
>    [-1, 3, BottleneckCSP, [1024, False]],  # 23 (P5/32-large)
> 
>    [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
>   ]
> ```
Have you achieved multi-task detection capabilities? Can you share how you do it?
Thank you.

**Comment by glenn-jocher** [comment]
@ahong007007 @thesby not sure I understand the question, YOLOv5 already uses multiple detection heads.

**Comment by ahong007007** [comment]
My current requirement is that head1 is used to detect dataset1, and head2 is used to detect dataset data sets. Because the data spaces of dataset1 and dataset2 are different (the label labeled by dataset1 may be a negative sample in dataset2, and the label category is not a containment relationship), it is hoped that different headn will be used for different detection tasks.

Could you please provide some suggestions for multi head, thank you

**Comment by ahong007007** [comment]
> @ahong007007 @thesby not sure I understand the question, YOLOv5 already uses multiple detection heads.

@glenn-jocher 
My current requirement is that head1 is used to detect dataset1, and head2 is used to detect dataset data sets. Because the data spaces of dataset1 and dataset2 are different (the label labeled by dataset1 may be a negative sample in dataset2, and the label category is not a containment relationship), it is hoped that different headn will be used for different detection tasks.

Could you please provide some suggestions for multi head, thank you

**Comment by glenn-jocher** [comment]
@ahong007007 data spaces of the different YOLOv5 heads are already different as they target different size objects.

You're free to modify the model and loss function etc to include however many heads you'd like.

**Comment by ahong007007** [comment]
@glenn-jocher thank you.

I want to modify the model like this:
```
backbone:
  # [from, number, module, args]
  [[-1, 1, Focus, [64, 3]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, BottleneckCSP, [128]],
   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
   [-1, 9, BottleneckCSP, [256]],
   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
   [-1, 9, BottleneckCSP, [512]],
   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
   [-1, 1, SPP, [1024, [5, 9, 13]]],
   [-1, 3, BottleneckCSP, [1024, False]],  # 9
  ]

# YOLOv5 head
head1:
  [[-1, 1, Conv, [512, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 6], 1, Concat, [1]],  # cat backbone P4
   [-1, 3, BottleneckCSP, [512, False]],  # 13

   [-1, 1, Conv, [256, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 4], 1, Concat, [1]],  # cat backbone P3
   [-1, 3, BottleneckCSP, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 14], 1, Concat, [1]],  # cat head P4
   [-1, 3, BottleneckCSP, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, BottleneckCSP, [1024, False]],  # 23 (P5/32-large)

   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  ]

head2:
  [[-1, 1, Conv, [512, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 6], 1, Concat, [1]],  # cat backbone P4
   [-1, 3, BottleneckCSP, [512, False]],  # 13

   [-1, 1, Conv, [256, 1, 1]],
   [-1, 1, nn.Upsample, [None, 2, 'nearest']],
   [[-1, 4], 1, Concat, [1]],  # cat backbone P3
   [-1, 3, BottleneckCSP, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 14], 1, Concat, [1]],  # cat head P4
   [-1, 3, BottleneckCSP, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, BottleneckCSP, [1024, False]],  # 23 (P5/32-large)

   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  ]
```
How to modify the parse_model function and the functions in yolo.py, how to increase the loss loss function, do you have any suggestions? thanks

**Comment by glenn-jocher** [comment]
@ahong007007 sure you can experiment with different architectures. I can't help you since we are very under-resourced as is, but maybe someone in the community can assist.
